### PR TITLE
Modernise the appearance of the plugin manager

### PR DESCRIPTION
**Pull request by janfaracik**
<details>
  <summary><b>📸  Screenshots</b></summary>
image
Screenshot 2021-11-12 at 19 29 22
<br><br><br>
image
Screenshot 2021-11-12 at 19 34 23
<br><br><br>
image
Screenshot 2021-11-12 at 19 29 56
<br><br><br>
image
Screenshot 2021-11-12 at 19 30 02
</details>

### Proposed changelog entries

* Modernise the appearance of the plugin manager

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by janfaracik** [comment]
⬇️ **Discussion around this would be good** ⬇️
I've added an `.editorconfig` file to ensure consistency in frontend files going forward, would be good to discuss this to see if theres a demand for this/any issues with it :)

File: core/src/main/resources/hudson/PluginManager/installed.jelly
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/installed.jelly:227]: 🚫 Blocker: You can now uninstall plugins that are depended on as mandatory dependeices

I can uninstall the credentials plugin which shouldn't be possible

**Review by timja — DISMISSED** [review]
spacing could probably be improved with banners?
image

Especially when there's 2 of them:

lots of spacing between (maybe too much and none under the text)
image

----

Blockers are the disabling and uninstalling of plugins that shouldn't be able to be.

**Comment by timja** [comment]
> ⬇️ **Discussion around this would be good** ⬇️ I've added an `.editorconfig` file to ensure consistency in frontend files going forward, would be good to discuss this to see if theres a demand for this/any issues with it :)

fine with me

**Comment by fqueiruga** [comment]
This is the PR I was waiting for!!! 🎉

I'm going to wait until you address Tim's feedback before taking a deep look into this. One suggestion is that you provide screenshots / video of the default theme. It's what most people will be using and the default.

File: war/src/main/less/base/spacing.less
Comment (fqueiruga) [inline war/src/main/less/base/spacing.less:3]: I'm going to suggest you go with the same sizing as established by bootstrap margin and padding utilities. That way those familiar with it will find it easier.

File: war/src/main/less/modules/page-header.less
Comment (fqueiruga) [inline war/src/main/less/modules/page-header.less:108]: Why converting from rem to px? rems should be preferred IMO

**Comment by fqueiruga** [comment]
Off topic, I just noticed the search input microinteraction. I love it!

https://user-images.githubusercontent.com/5738588/141614238-4346d861-d4de-45c1-8aa2-8e051bc98ff7.mov

**Comment by timja** [comment]
you can no longer un-install plugins that aren't allowed to be, but you can still disable them

**Review by uhafner — COMMENTED** [review]
Thanks for starting yet again another improvement! The changes look very nice in general!

I don't know how far you want to go here, so I hope you don't mind that I also add some comments that discuss some of the usability problems in the existing dialog (that might be out of scope of this pull request).

# General
- The tabs are using different URLs (pages), that means I cannot install and update plugins in a single step. When I move to another tab, then my selections are deleted.
- All tabs: some have a filter, some not. It makes sense to move the filter inside the tab so the tabs are not moving vertically anymore: I think only the table should move, the tabs should have a fixed position. This would also move the filter button more close to the corresponding table.
# Installed
- The installed version column seems to be so far away from the version that is going to be installed. But I have no idea how to improve that. Maybe changing the order of release date and version?
- It would be helpful if each table row would be expandable (see jQuery DataTables) so one could see the release notes (if available).
- I would move the "Update information obtained: 4 hr 11 min ago" text to the table right side and replace the button with a non-primary button. (Or even better a button with a "Reload" button and a corresponding reload icon) 
- The download button should be simply called "Download" and should be the only primary one. 
- The Buttons Select: "All, Compatible, None" should be actually buttons and should be placed before the download button. Normally, such a "All/None" button is shown inside the table header with a checkbox so one can select all by clicking the checkbox and select none by deactivating the checkbox. I know from a lot of people that they are not aware that it is possible to select all of the plugins. 
- Shaded but selectable rows are in progress or failed: this should be named "Shaded rows are in progress or failed": I cannot select anything if the plugin has been installed.

# Available 
- The boxes are really nice now. I wonder if it would be possible to show the tags "warning", "deprecated" and "adoption" as single buttons in the style of the content tags (like "github" or "pipeline") without the boring and repeating texts. They can be placed right after the tags with the corresponding color. I think for me it would be sufficient if the corresponding texts are shown as a tooltip. Otherwise a lot of place is used for the repeated and boring texts "This plugin is up for adoption! We are looking for new maintainers. Visit our Adopt a Plugin initiative for more information.". When these things are also shown as tags, then one can filter by them as well
- Install and Download buttons: these buttons should be replaced by a checkbox ([ ] install without restart) and a "Download" button that considers the checkbox on how to actually install. Also only a single word should be used as text of the button. 
- I would move the "Update information obtained: 4 hr 11 min ago" text to the table right side and replace the button with a non-primary button. (Or even better a button with a "Reload" button and a corresponding reload icon) 

# Installed
- The new downgrade button looks awesome. I would recommend to add an additional button for the upgrade. As user it makes no sense that I need to switch to another tab to show the updates. Maybe there should be a toggle to show only plugins with updates.
- The buttons are very far way from the text on my screen. I would suggest to place them before the text as first column:
Bildschirmfoto 2021-11-13 um 16 03 36
- The adoption and warning boxes should be replaced with single tags (see above)
- The "Restart" button at the button is not visible (on the available tab it is?). I think it would be helpful if it is sticky. And it should be a destructive one? Again: I am not a fan of using buttons with sentences, they should be short as possible. 

# Advanced
- I mentioned that already in another PR: it would make sense to have one one thing with one primary button on one tab. It would be better to have more tabs.

**Comment by timja** [comment]
Blockers are addressed now (other than tests), although there's a fair bit of feedback WDYT @janfaracik?

**Comment by janfaracik** [comment]
> Blockers are addressed now (other than tests), although there's a fair bit of feedback WDYT @janfaracik?

There's definitely some good suggestions in here I'd like to address + update the MR with - I'll comment again when it should be good for another look at :)

File: core/src/main/resources/lib/layout/pageHeader.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/pageHeader.jelly:28]: revert?

File: core/src/main/resources/lib/layout/layout.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/layout.jelly:223]: I wanted to move the app-bar (the title/controls row) to `layout.jelly` to make it easier to use in the future, so that devs wouldn't need to add all the `div` structuring for it. 

This way from a page all you need to do is:

```
<l:app-bar title="${%Plugin Manager}">
      <div class="jenkins-search">
        <input
          class="jenkins-search__input"
          type="search"
          id="filter-box"
          value="${request.getParameter('filter')}"
          placeholder="${filtered == 'true' ? '%Search' : '%Filter'}"/>
        <div class="jenkins-search__icon">
          <l:ionicon name="search-outline"/>
        </div>
      </div>
    </l:app-bar>
```

to use it. Hope that answers your question?

File: core/src/main/resources/lib/layout/pageHeader.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/pageHeader.jelly]: Good spot :)

**Comment by janfaracik** [comment]
I've done some cleanup on this PR/formatted some code correctly/fixed a few issues. There's definitely some great ideas listed above that I'd like to have a crack at, but I think another branch/PR would be best suited for that. 

It'd be great if people could take another look over this PR and make sure everything is ok :)

**Comment by daniel-beck** [comment]
This is probably coming too late, but I just saw this in a notification diff GitHub offered: do we want `l:ionicon` to be part of the core API forever? That seems to unnecessarily expose an implementation detail - the icon set/style - to callers?

**Comment by timja** [comment]
> This is probably coming too late, but I just saw this in a notification diff GitHub offered: do we want `l:ionicon` to be part of the core API forever? That seems to unnecessarily expose an implementation detail - the icon set/style - to callers?
> 
> Isn't that as if there's now a ton of code in core and plugins that explicitly referenced "tango"? That would be weird too.

It's not used outside of core yet, so safe to change (not in this PR though).
The API needs iterating for it anyway at some point. You can't put them in the sidebar or manage page yet as it's not integrated with `Icon` (outside of living in the class).

**Comment by timja** [comment]
> I've done some cleanup on this PR/formatted some code correctly/fixed a few issues. There's definitely some great ideas listed above that I'd like to have a crack at, but I think another branch/PR would be best suited for that.
> 
> It'd be great if people could take another look over this PR and make sure everything is ok :)

We also agreed this on UX sig meeting today.

Pending successful bom and ATH we'll look to ship this (unless any blocking feedback)

ATH started here:
https://github.com/jenkinsci/acceptance-test-harness/pull/716

Waiting on incremental build for bom

**Review by timja — APPROVED** [review]
Bom testing has passed: https://github.com/jenkinsci/bom/pull/756

ATH was tested in https://github.com/jenkinsci/acceptance-test-harness/pull/716.

and issues are resolved in https://github.com/jenkinsci/acceptance-test-harness/pull/718

I think we're ready to ship this.

Thanks!

File: .editorconfig
Comment (jetersen) [inline .editorconfig:1]: Since this is in the root of the repo. You ought to set `root = true`

https://editorconfig.org/#file-location
> When opening a file, EditorConfig plugins look for a file named .editorconfig in the directory of the opened file and in every parent directory. A search for .editorconfig files will stop if the root filepath is reached or an EditorConfig file with root=true is found.

```suggestion
root = true

[*.{jelly, js, less, css, hbs}]
```

**Review by jetersen — COMMENTED** [review]
Adding an editorconfig is great, but please add the `root = true`, so that editor will stop traversing all directories.

**Comment by jetersen** [comment]
@timja I have no further input. I do not think I can approve the frontend part of it 😅 At least it is looking great, so keep up the good work!

**Review by uhafner — APPROVED** [review]
Well, not much has changed since my last review. The new UI looks much better in most places. I think the deprecated banners are still too close to the text, but that was a problem before.

One thing that I really don't like is the position of the search bar. It is far too wide in the upper right corner (where it looks unrelated) and a lot of users will not find them when opening the new page the first time. It already is not ideal in the current UI (where it is above the tabs rather than right above the table), but now I think the usability is getting worse. 

Bildschirmfoto 2021-12-12 um 23 06 01

But anyway, we can improve that in followup PRs.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by uhafner** [comment]
I created a followup Jira so the left review comments will not be forgotten: https://issues.jenkins.io/browse/JENKINS-67366.

**Comment by janfaracik** [comment]
> I created a followup Jira so the remaining review comments will not be forgotten: https://issues.jenkins.io/browse/JENKINS-67366.

Awesome! I've got a follow up branch in the works to address a lot of your comments - thanks :)

**Comment by Bananeweizen** [comment]
The downgrade buttons are not aligned. Their position depends on the text width of the version number.
image

**Comment by timja** [comment]
can you create an issue on issues.jenkins.io please and link here?

so that someone can take a look when they have time

**Comment by basil** [comment]
@janfaracik Causes [JENKINS-68277](https://issues.jenkins.io/browse/JENKINS-68277).

**Comment by basil** [comment]
Causes [JENKINS-68291](https://issues.jenkins.io/browse/JENKINS-68291).

**Comment by daniel-beck** [comment]
What's the purpose of editorconfig when the most important JS file in the project isn't following this to begin with? I currently have to fight against editorconfig on every single line I edit in `hudson-behavior.js` unless I want mixed-width indentation in the file.

**Comment by timja** [comment]
> What's the purpose of editorconfig when the most important JS file in the project isn't following this to begin with? I currently have to fight against editorconfig on every single line I edit in `hudson-behavior.js` unless I want mixed-width indentation in the file.

Filed https://github.com/jenkinsci/jenkins/pull/6863

**Comment by daniel-beck** [comment]
This editorconfig continues to be incredibly annoying, since not all Jelly files are currently indented using 2 spaces.

**Pull request merged**
